bulldozer/signals.go

```diff
@@ -21,109 +21,389 @@ import (
 	"strings"
 
 	"github.com/palantir/bulldozer/pull"
+	"github.com/pkg/errors"
 	"github.com/rs/zerolog"
 )
 
+type Signal []string
+
 type Signals struct {
-	Labels            []string `yaml:"labels"`
-	CommentSubstrings []string `yaml:"comment_substrings"`
-	Comments          []string `yaml:"comments"`
-	PRBodySubstrings  []string `yaml:"pr_body_substrings"`
-	Branches          []string `yaml:"branches"`
-	BranchPatterns    []string `yaml:"branch_patterns"`
+	Labels            Signal `yaml:"labels"`
+	CommentSubstrings Signal `yaml:"comment_substrings"`
+	Comments          Signal `yaml:"comments"`
+	PRBodySubstrings  Signal `yaml:"pr_body_substrings"`
+	Branches          Signal `yaml:"branches"`
+	BranchPatterns    Signal `yaml:"branch_patterns"`
+	MaxCommits        int `yaml:"max_commits"`
 }
 
-func (s *Signals) Enabled() bool {
+type SignalsMatches struct {
+	Signals
+}
+
+func (s *Signals) size() int {
 	size := 0
 	size += len(s.Labels)
 	size += len(s.CommentSubstrings)
 	size += len(s.Comments)
 	size += len(s.PRBodySubstrings)
 	size += len(s.Branches)
 	size += len(s.BranchPatterns)
-	return size > 0
+	if s.MaxCommits > 0 {
+		size += 1 
+	}
+	return size
 }
 
-// Matches returns true if the pull request meets one or more signals. It also
+func (s *Signals) Enabled() bool {
+	return s.size() > 0
+}
+
+// MatchesAll returns true if the pull request matches ALL of the signals. It also
+// returns a description of the match status. The tag argument appears
+// in this description and indicates the behavior (trigger, ignore) this
+// set of signals is associated with.
+func (s *Signals) MatchesAll(ctx context.Context, pullCtx pull.Context, tag string) (bool, string, error) {
+	matches := SignalsMatches{}
+	var err error
+
+	_, matches.Labels, err = s.labelMatches(ctx, pullCtx, tag)
+	// TODO: Collect errors as a list
+	if err != nil {
+		return false, "", err
+	}
+
+	_, matches.Comments, err = s.commentMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	_, matches.CommentSubstrings, err = s.commentSubstringMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	_, matches.PRBodySubstrings, err = s.prBodyMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	_, matches.Branches, err = s.branchMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	_, matches.BranchPatterns, err = s.branchPatternMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	_, matches.MaxCommits, err = s.maxCommitsMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	logger := zerolog.Ctx(ctx)
+	logger.Debug().Msgf("matches count: %d", matches.size())
+	logger.Debug().Msgf("signals count: %d", s.size())
```

Comment:
These should probably be a single message so they stick together.

Reply:
Removed, used for debugging